Support function update

glyn opened the pull request:
Fixes https://github.com/projectriff/function-controller/issues/9

glyn committed:
Support function update

Fixes https://github.com/projectriff/function-controller/issues/9 and makes
https://github.com/projectriff/function-controller/issues/5 much less likely
to happen.

ericbottard commented on pkg/controller/controller.go:
```diff
@@ -163,6 +172,27 @@ func (c *ctrl) onFunctionAddedOrUpdated(function *v1.Function) {
 }
 
+func (c *ctrl) onFunctionUpdated(oldFn *v1.Function, newFn *v1.Function) {
+	if oldFn.Name != newFn.Name {
```
I'm pretty sure this can't happen, per design, no? An "update" for k8s only happens if you're changing the same object, which is identified by its name+namespace

glyn replied:
Agreed. It's really an assertion since the behaviour is so predicated on the name and the namespace not changing.

ericbottard commented on pkg/controller/controller.go:
```diff
@@ -95,6 +96,12 @@ type topicKey struct {
 }
 
+// A Δfn represents a pair of functions involved in an update
+type Δfn struct {
```
How on earth do I type a Δ? Although I find it pretty cool, I think this doesn't work in practice. We need to think about users with a non US keyboard layout

glyn replied:
On Mac, it's alt+J. It's not US keyboard specific - I have a UK keyboard. ;-)

I expected this reaction BTW, but I was encouraged by Onsi's use of Ω (alt+Z) in Gomega. :-)

If you still feel uncomfortable, I'l refactor `∆fn` to `deltaFn`.

ericbottard replied:
I looked it up, and I tried it. It is US/UK specific, believe me, it doesn't work on a french layout (it does when I switch to US). I really think we should switch, sadly

glyn replied:
@ericbottard Actually, I've repented. I'll push an update soon.

ericbottard commented on pkg/controller/deployer.go:
```diff
@@ -125,6 +128,25 @@ func (d *deployer) Undeploy(function *v1.Function) error {
 }
 
+func (d *deployer) Update(function *v1.Function) error {
+	deployment := v1beta1.Deployment{
```
I had a working version that did a get + update.
As a matter of fact, I'm not even sure we want/need a new `Update` function. `Deploy` with an `if` should be fine

glyn replied:
I'd prefer to _mirror_ the structure of the k8s API and have separate `Deploy` and `Update` operations rather than overload functions to do multiple things.

ericbottard replied:
OK, fair enough. I think the no of replicas should not be set to zero then, at the very least

glyn committed:
Changes from review

Set the updated deployment to use the current actual number of replicas.

Update the functions map. Note: this doesn't currently have any effect since
the only fields of the map value that are used are the function name and
namespace.

glyn commented:
@ericbottard Changes added in a new commit.

glyn commented:
@ericbottard The second commit sets the updated deployment to have the current actual number of replicas rather than defaulting to 0, as we discussed.

This, combined with riff's scaling behaviour, _sometimes_ prevents k8s from completing the rolling update of the function pod until a quiet period when riff scales the pod to 0. For 0.0.4, I'm inclined to switch back to 0 replicas in the updated deployment and log an issue to explore this further.

Do you agree with that approach? Given the tight timescale, I'll add a third commit which does that and you can take it or leave it. :-)

glyn committed:
On function update, set updated deployment to have 0 replicas

See https://github.com/projectriff/function-controller/issues/17

glyn commented:
@ericbottard 3rd commit available for review/testing.

ericbottard commented:
Merged with all 3 commits squashed as 3afb01f
